test/Microsoft.OpenApi.Tests/Workspaces/OpenApiFragmentsTests.cs

```diff
@@ -0,0 +1,254 @@
+        };
+
+
+        private static readonly OpenApiComponents _componentsFragment = new OpenApiComponents
+        {
+            Schemas = { ["schema1"] = _schemaFragment },
+            Responses = { ["response1"] = _responseFragment },
+            Parameters = { ["parameter1"] = _parameterFragment },
+            Examples = { ["example1"] = _exampleFragment },
+            RequestBodies = { ["requestBody1"] = _requestBodyFragment },
+            Headers = { ["header1"] = _headerFragment },
+            SecuritySchemes = { ["securityScheme1"] = _securitySchemeFragment },
+            Links = { ["link1"] = _linkFragment },
+            Callbacks = { ["callback1"] = _callbackFragment }
+        };
+
+        public static IEnumerable<object[]> ElementsWhichCannotBeReferenced =>
+            new List<object[]>
+            {
+                new object[] { new OpenApiArray() },
+                new object[] { new OpenApiBinary(new byte[0]) },
+                new object[] { new OpenApiBoolean(true) },
+                new object[] { new OpenApiByte(1) },
+                new object[] { new OpenApiDate(DateTime.Now) },
+                new object[] { new OpenApiDateTime(DateTime.Now) },
+                new object[] { new OpenApiDouble(1D) },
+                new object[] { new OpenApiFloat(1F) },
+                new object[] { new OpenApiInteger(1) },
+                new object[] { new OpenApiLong(1) },
+                new object[] { new OpenApiNull() },
+                new object[] { new OpenApiObject() },
+                new object[] { new OpenApiPassword("pw123") },
+                new object[] { new OpenApiString("string123") },
+                new object[] { new OpenApiComponents() },
+                new object[] { new OpenApiContact() },
+                new object[] { new OpenApiDiscriminator() },
+                new object[] { new OpenApiEncoding() },
+                new object[] { new OpenApiExternalDocs() },
+                new object[] { new OpenApiInfo() },
+                new object[] { new OpenApiLicense() },
+                new object[] { new OpenApiMediaType() },
+                new object[] { new OpenApiOAuthFlow() },
+                new object[] { new OpenApiOAuthFlows() },
+                new object[] { new OpenApiOperation() },
+                new object[] { new OpenApiPathItem() },
+                new object[] { new OpenApiReference() },
+                new object[] { new OpenApiSecurityRequirement() },
+                new object[] { new OpenApiServer() },
+                new object[] { new OpenApiServerVariable() },
+                new object[] { new OpenApiXml() }
+            };
+
+        [Theory]
+        [MemberData(nameof(ElementsWhichCannotBeReferenced))]
+        public void ElementsWhichCannotBeReferencedShouldThrow(IOpenApiElement element)
+        {
+            // Arrange & Act
+            Action resolveReference = () => element.ResolveReference(new JsonPointer("/"));
```

Comment:
Rather than adding ResolveReference to IOpenApiElement but then checking to make sure that any non-referenceable  element just throws, could we instead move ResolveReference into IOpenApiReferencable?